week4/hw-week4.playground/Contents.swift

```diff
@@ -0,0 +1,165 @@
+import UIKit
+
+/*
+ a) In the assignment for Week 3, part D asked you to write a function that would compute the average of an array of Int. Using that function and the array created in part A, create two overloaded functions of the function average.
+ */
+
+
+
+func average(val: [Int]?){
+    var a1:[Int]? = val
+    var sum=0
+    guard let val = a1 else {
+        print("The array is nil. Calculating the average is impossible.")
+        return
+    }
+    
+    for i in val{
+        sum = sum+i
+    }
+    
+    print("The average of the values in the array is \(Double(sum/val.count))")
+}
+
+func average(val: [Double]?){
+    var a1:[Double]? = val
+    var sum=0.0
+    guard let val = a1 else {
+        print("The array is nil. Calculating the average is impossible.")
+        return
+    }
+    
+    for i in val{
+        sum = sum+i
+    }
+    
+    print("The average of the values in the array is \(Double(sum/Double(val.count)))")
+}
+
+func average(val: [Float]?){
+    var a1:[Float]? = val
+    var sum=Float(0)
+    guard let val = a1 else {
+        print("The array is nil. Calculating the average is impossible.")
+        return
+    }
+    
+    for i in val{
+        sum = sum+i
+    }
+    
+    print("The average of the values in the array is \(Float(sum/Float(val.count)))")
+}
+
+var optArray:[Int]? = [10,20,30]
+var optDoubleArray:[Double]? = [10.5,20.5,30.5]
+var optFloatArray:[Float]? = [10.5,20.5,30.5]
+
+
+average(val: optArray)
+average(val: optDoubleArray)
+average(val: optFloatArray)
+
+/*
+ b) Create an enum called Animal that has at least five animals. Next, make a function called theSoundMadeBy that has a parameter of type Animal. This function should output the sound that the animal makes. For example, if the Animal passed is a cow, the function should output, “A cow goes moooo.”
+ Hint: Do not use if statements to complete this section.
+ Call the function twice, sending a different Animal each time.
+ */
+
+enum Animals: CaseIterable {
+    case cow, dog, pig, sheep, elephant
+}
+
+func theSoundMadeBy(animal: Animals)-> String{
+    switch animal {
+    case .cow:
+        return ("A cow goes moooo.")
+    case .dog:
+        return ("A dog goes bow.")
+    case .elephant:
+        return ("A elephant goes trumpeting")
+    case .pig:
+        return ("A pig goes snort")
+    case .sheep:
+        return ("A sheep goes baa")
+    }
+}
+
+for animal in Animals.allCases{
+    print(theSoundMadeBy(animal: animal))
+}
+
```

Comment:
Nice Job using caseIterable for the loop. 

This code was simple and direct. 

A refactor idea would be making the enum have a case and the function only then prints the rawValue of the case rather than having to know all the sounds to print.
In this small of a dataset it's negligible, but if the values of the enum ever change, you could have a function that doesn't contain all the case options. 

for example: 
enum Animals2: String, CaseIterable {
	case turtles = "cowabunga", dog = "Woof"
}
func sound(animal: Animals2) -> String {
	return animal.rawValue
}

for animal in Animals2.allCases {
	print(sound(animal: animal))
}